roachprod-microbench: post GitHub issues for performance regressions

ibreakthecloud opened the pull request:
Previously, performance regressions detected during the weekly microbenchmark comparison were only reported via Slack notifications. This made it difficult to track and ensure timely follow-up on regressions, as they were often discussed informally without formal issue tracking.

This change extends the existing `--post-issues` flag to work with the compare command. When enabled, the system automatically creates GitHub issues for performance regressions that exceed 20% (the "red" regression threshold). Each issue includes:
- Package name and list of regressed benchmarks
- Regression percentages and formatted deltas
- Link to the Google Sheet with detailed comparison data
- Labels: O-microbench and C-performance for easy filtering

The implementation reuses the same GitHub posting infrastructure and environment variables (GITHUB_BRANCH, GITHUB_SHA, GITHUB_BINARY) as the existing benchmark failure reporting. Issues are created per package to avoid spam, with up to 10 regressions listed in each issue summary.

[Example GitHub Issue](https://github.com/cockroachdb/exalate-test/issues/222) 
screenshot:
<img width="1723" height="662" alt="image" src="https://github.com/user-attachments/assets/27d4cb6f-c3df-42d6-b75c-fd533c4656d8" />


Ticket: https://cockroachlabs.atlassian.net/browse/CODESYS-106

Epic: None
Release note: None

Commit:
roachprod-microbench: post GitHub issues for performance regressions

Previously, performance regressions detected during the weekly microbenchmark
comparison were only reported via Slack notifications. This made it difficult
to track and ensure timely follow-up on regressions, as they were often
discussed informally without formal issue tracking.

This change extends the existing `--post-issues` flag to work with the compare
command. When enabled, the system automatically creates GitHub issues for
performance regressions that exceed 20% (the "red" regression threshold). Each
issue includes:
- Package name and list of regressed benchmarks
- Regression percentages and formatted deltas
- Link to the Google Sheet with detailed comparison data
- Labels: O-microbench and C-performance for easy filtering

The implementation reuses the same GitHub posting infrastructure and
environment variables (GITHUB_BRANCH, GITHUB_SHA, GITHUB_BINARY) as the
existing benchmark failure reporting. Issues are created per package to avoid
spam, with up to 10 regressions listed in each issue summary.

This change also renames `postBenchmarkIssue` to `postIssuesToGitHub` for
consistency, as it now handles both execution failures and performance
regressions.

Epic: None
Release note: None

rishabh7m commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -207,6 +207,7 @@ if [ -d "$output_dir/experiment" ] && [ "$(ls -A "$output_dir/experiment")" ] \
     --sheet-desc="$sheet_description" \
     ${influx_token:+--influx-token="$influx_token"} \
     ${influx_host:+--influx-host="$influx_host"} \
+    ${TRIGGERED_BUILD:+--post-issues} \
```
Why this is required and cannot be the default setting? Is there any similar flag available for the slack notification?

ibreakthecloud replied:
This flag already exists, I am just using it. See [here](https://github.com/ibreakthecloud/cockroach/blob/f71f5f2f4e72e0cbe6f2d94ba97d7abee9776e8b/pkg/cmd/roachprod-microbench/main.go#L119).

rishabh7m reviewed (changes requested):
- How was this tested? Can you paste a screenshot or the link of any generated issue?
- Does this change needs to be backported?

ibreakthecloud commented on pkg/cmd/roachprod-microbench/compare.go:
```diff
@@ -349,6 +351,52 @@ func (c *compare) compareUsingThreshold(comparisonResultsMap model.ComparisonRes
+			formatter, req := createRegressionPostRequest(pkgName, regressions, sheetLink, c.sheetDesc)
+			err := postIssuesToGitHub(c.ctx, l, formatter, req)
+			if err != nil {
+				return errors.Wrapf(err, "failed to post regression issue for package %s", pkgName)
```
Fair point. I will log the error and continue.

ibreakthecloud commented:
@rishabh7m 
> How was this tested? Can you paste a screenshot or the link of any generated issue?

No, it was not, let me test and update this PR.

> Does this change needs to be backported?

No I don't think so.

ibreakthecloud commented:
@srosenberg 

> How was this tested? Can you paste a screenshot or the link of any generated issue?
>> I've added the unit test that verifies the format of issue posted during regression.
>>> The unit test is great, but it would still be nice to test it end-to-end. You can create a dummy issue to serve as an example.

Fair point, I will create a dummy issue and update the description.


> In case of a misconfiguration (or other bug), what if every package results in a regression? We should limit the total (possible) number of created GH issues.

I will limit it to 5 issues.

herkolategan reviewed (changes requested):
I would like to understand which team labels will be applied to each `package` regression issue. Since this adds the `release-blocker` label, it will become the duty of those team(s) to address or close this issue. Although an issue could end up containing benchmarks owned by multiple teams.

nameisbhaskar commented on pkg/cmd/roachprod-microbench/compare.go:
```diff
@@ -349,6 +351,53 @@ func (c *compare) compareUsingThreshold(comparisonResultsMap model.ComparisonRes
+func (c *compare) postRegressionIssues(
+	links map[string]string, comparisonResultsMap model.ComparisonResultsMap,
+) error {
+	loggerCfg := logger.Config{Stdout: os.Stdout, Stderr: os.Stderr}
```
Given that the code uses "log.Printf", we may just use that.

ibreakthecloud commented:
@nameisbhaskar I've addressed the review comments.

nameisbhaskar reviewed (approved):
LGTM!

srosenberg commented on pkg/cmd/roachprod-microbench/github.go:
```diff
@@ -43,6 +43,47 @@ func createBenchmarkPostRequest(
+
+	sb.WriteString(fmt.Sprintf("\nDetailed comparison: %s\n", sheetLink))
+
+	benchmarkName := regressions[0].benchmarkName
```
See my comment about finding the correct owner [1].

[1] https://github.com/cockroachdb/cockroach/pull/157045#issuecomment-3649977842

srosenberg reviewed (changes requested):
We need to address the following,

- find and assign correct owner, not the catch-all `test-eng`
- describe what branches and SHAs are being compared

ibreakthecloud commented on pkg/cmd/roachprod-microbench/main.go:
```diff
@@ -173,6 +173,12 @@ func makeCompareCommand() *cobra.Command {
+			if c.postIssues {
+				err = c.postRegressionIssues(links, comparisonResult)
+				if err != nil {
+					return err
```
I think calling `pushToInfluxDB`, moment we have the `comparisonResult` would be better. I'll move it up LMK if you have different thoughts.

ibreakthecloud replied:
I've updated the code to use an error accumulation pattern. Now all operations (Google Sheets, Slack,
  GitHub issues, and InfluxDB) are attempted regardless of individual failures, and errors are collected and
  propagated at the end.

  This ensures:
  - All operations are attempted
  - Also, errors are not swallowed (CI will fail if any operation fails)

srosenberg commented on pkg/cmd/roachprod-microbench/compare.go:
```diff
@@ -349,6 +349,49 @@ func (c *compare) compareUsingThreshold(comparisonResultsMap model.ComparisonRes
+
+		if len(regressions) > 0 {
+			sheetLink := links[pkgName]
+			formatter, req := createRegressionPostRequest(pkgName, regressions, sheetLink, c.sheetDesc)
```
> have one more flag, --version-context="v24.1.0 -> v24.2.0" which will be used by both sheets and Github. And rename --sheet-desc to generate-sheet (BoolVar)

Seems like the cleanest option.

ibreakthecloud replied:
Updated the code with this approach. Also update the issue screenshot & URL (it now points to same github issue but comment) in PR body.

srosenberg replied:
I'm not seeing `--version-context`; did I miss the commit for that? Also, the [URL](https://github.com/crltest-issue-syncing/cockroach/issues/5583#issuecomment-3664039669) is not accessible to me, I get `404`.

ibreakthecloud replied:
> I'm not seeing --version-context

Ah! My bad. Can you check again?

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.97m ±0% | 11.99m ±0% | ~ | p=0.074 n=15    |
| ⚪ **allocs/op** | 8.159k ±0% | 8.161k ±0% | ~ | p=0.943 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/0b4f4b8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/0b4f4b844075cd77cecf4166abe4312482a0fe17/bin/pkg_sql_tests benchdiff/0b4f4b8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/0b4f4b8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/9918045/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/99180451dea5c745ea9e5c86e60aeb125491be4d/bin/pkg_sql_tests benchdiff/9918045/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/9918045/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=9918045 --new=0b4f4b8 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.453m ±1% | 3.449m ±1% | ~ | p=0.595 n=15    |
| ⚪ **allocs/op** | 2.082k ±0% | 2.081k ±0% | ~ | p=0.276 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/0b4f4b8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/0b4f4b844075cd77cecf4166abe4312482a0fe17/bin/pkg_sql_tests benchdiff/0b4f4b8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/0b4f4b8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/9918045/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/99180451dea5c745ea9e5c86e60aeb125491be4d/bin/pkg_sql_tests benchdiff/9918045/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/9918045/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=9918045 --new=0b4f4b8 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.750m ±1% | 3.756m ±1% | ~ | p=0.713 n=15    |
| ⚪ **allocs/op** | 4.178k ±0% | 4.178k ±0% | ~ | p=0.943 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/0b4f4b8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/0b4f4b844075cd77cecf4166abe4312482a0fe17/bin/pkg_sql_tests benchdiff/0b4f4b8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/0b4f4b8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/9918045/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/99180451dea5c745ea9e5c86e60aeb125491be4d/bin/pkg_sql_tests benchdiff/9918045/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/9918045/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=9918045 --new=0b4f4b8 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/0b4f4b844075cd77cecf4166abe4312482a0fe17/20359493162-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/99180451dea5c745ea9e5c86e60aeb125491be4d/20359493162-1/\* old/
```

</details>

_built with commit: [0b4f4b844075cd77cecf4166abe4312482a0fe17](https://github.com/cockroachdb/cockroach/commit/0b4f4b844075cd77cecf4166abe4312482a0fe17)_

<!-- microbench-ci -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.29m ±1% | 11.32m ±1% | ~ | p=0.267 n=15    |
| ⚪ **allocs/op** | 8.162k ±1% | 8.159k ±1% | ~ | p=0.911 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/b500279/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/b500279d1c4ef18ab82963a57fe1847a789969f4/bin/pkg_sql_tests benchdiff/b500279/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/b500279/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/43add13/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/43add13db8696097b67ab8de556bf0a69431dea6/bin/pkg_sql_tests benchdiff/43add13/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/43add13/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=43add13 --new=b500279 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.251m ±0% | 3.277m ±0% | +0.79% | p=0.000 n=15    |
| ⚪ **allocs/op** | 2.082k ±0% | 2.082k ±0% | ~ | p=0.385 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/b500279/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/b500279d1c4ef18ab82963a57fe1847a789969f4/bin/pkg_sql_tests benchdiff/b500279/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/b500279/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/43add13/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/43add13db8696097b67ab8de556bf0a69431dea6/bin/pkg_sql_tests benchdiff/43add13/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/43add13/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=43add13 --new=b500279 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.628m ±2% | 3.617m ±2% | ~ | p=0.967 n=15    |
| ⚪ **allocs/op** | 4.187k ±0% | 4.190k ±0% | ~ | p=0.630 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/b500279/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/b500279d1c4ef18ab82963a57fe1847a789969f4/bin/pkg_sql_tests benchdiff/b500279/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/b500279/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/43add13/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/43add13db8696097b67ab8de556bf0a69431dea6/bin/pkg_sql_tests benchdiff/43add13/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/43add13/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=43add13 --new=b500279 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/b500279d1c4ef18ab82963a57fe1847a789969f4/20366155031-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/43add13db8696097b67ab8de556bf0a69431dea6/20366155031-1/\* old/
```

</details>

_built with commit: [b500279d1c4ef18ab82963a57fe1847a789969f4](https://github.com/cockroachdb/cockroach/commit/b500279d1c4ef18ab82963a57fe1847a789969f4)_

<!-- microbench-ci -->

srosenberg reviewed (approved):
This is looking better! I would suggest a couple of small improvements to the formatting of the issue; otherwise, LGTM.

In the linked example, we could further simplify the title to make it canonical like so,

```
pkg/backup: potential performance regression 
```

The first part of the `Comparison` string can be mapped into an existing GH label, i.e., `branch-release-25.3`. (Since the body already contains the `Comparison` string, it seems redundant to have it in the title; the issue label reflects the release branch that is being compared to `master`.)

I also wonder if it's easy to generate a direct link to the [dashboard](https://microbench.testeng.crdb.io/dashboard/) similar to how it's done in the spreadsheet. 

CC @herkolategan for his opinion.

herkolategan reviewed (approved):
Thanks for making the changes, LGTM!

I will note two things that won't be part of this PR, but something we should note:

- There are some noisy benchmarks that will continuously cause posts.
- We currently still use `n2-standard-32` machines, we should switch to the `c4` machines to improve SNR.

ibreakthecloud commented:
TFTR @srosenberg @herkolategan 
> This is looking better! I would suggest a couple of small improvements to the formatting of the issue; otherwise, LGTM.
>
>In the linked example, we could further simplify the title to make it canonical like so,
>
>pkg/backup: potential performance regression 
>The first part of the Comparison string can be mapped into an existing GH label, i.e., branch-release-25.3. (Since the body >already contains the Comparison string, it seems redundant to have it in the title; the issue label reflects the release branch >that is being compared to master.)
>
>I also wonder if it's easy to generate a direct link to the [dashboard](https://microbench.testeng.crdb.io/dashboard/) similar >to how it's done in the spreadsheet.

I've update the code with following changes:
- Simplified Title, e.g. "pkg/storage: potential performance regression"
- Extracting version from version_context and applying label, e.g. for this (v25.3.4 (1cddb3a2) -> refs/heads/master (ddeef368)), label will be `branch-release-25.3`
- Added dashboard link for each benchmark.

Issue and screenshot in PR reflects the latest changes.

I'm merging this PR, feel free to create new ticket if any further enhancement is needed.

ibreakthecloud commented:
bors r+
